RFD 0087: Access request notification routing (#16288)
gravitational/teleport: rfd/0087-access-request-notification-routing.md

---
authors: Hugo Hervieux (hugo.hervieux@goteleport.com)
state: draft
---
# RFD 87 - Access request notification routing

## Required Approvers

* Engineering r0mant && ???
* Security @reedloden
* Product: (@xinding33 || @klizhentas)

## What

Provide a granular way for administrators to configure which access-requests notifications should be delivered to whom
over which medium.

## Why

- Teleport access request routing is not configured the same way depending on the plugin: Pagerduty uses annotations
  while most plugins use a role_to_recipient map in the plugin config. Users leveraging both PagerDuty and Slack plugins 
  have to deal with two configuration languages;
- Current Pagerduty request routing applies the same rule to all requests from the same role, regardless of the
  requested role. All access requests don't have the same severity, users wanting to route access requests differently
  have to create multiple roles. Depending on the company size and structure cardinality can become a problem.

>   See https://github.com/gravitational/teleport-plugins/issues/597 ; and

Author: obscure Garner bullet-list syntax, will switch to a simpler one

- `plugin` is a string describing which plugin should pick up this target;
  By default each plugin are listening for their own name (`slack` for the slack plugin,
  `msteams` for the Microsoft Teams one, ...). But the plugin name can be configurable
  in the plugin configuration. This allows support for multiple instances of the same
  plugin (e.g. many orgs have multiple slack workspaces, they can run 1 plugin per workspace); and
- `recipient` is a string describing the access request recipient. Different kinds of recipients
  are supported depending on the plugin (user name, channel name, service name, user id,
  channel id, URL, ...).

### New Resource: `access_routing_rule`

This proposal introduces a new resource to add routing information to newly created access-requests.

The motivation behind adding a new resource, opposed to add new fields in `role` is to
limit the `role` spec inflation and allow more flexibility to create role-agnostic routing rules.

Routing rules are evaluated on access-request creation and will add elements to a new field in
access-requests: `targets`.

An example access_routing_rule would look like:

```yaml

> kind: access_routing_rule

Reviewer: Nit: Not sure if it's too verbose but `access_request_routing_rule` would probably be a bit more accurate name for this resource.

Author: done

version: v1
metadata:
  name: example
spec:
  targets:
  # simplified syntax

>   - condition: 'equals(resource.spec.requested_role, "prod-rw")'

Reviewer: What is the resource that this condition will be evaluated on? Access request?

>     recipient: "Teleport Alice"

Reviewer: Can this be a list?

    plugin: "pagerduty"
  # raw expression
  - expression: >
      ifelse(
        equals(resource.spec.requested_role, "prod-rw"),
        pair(),
        pair("msteams", "alice@example.com") 
      )
```

Invoked with the role `"prod-rw"`, this would produce the following `access-request`

```yaml
kind: access_request
spec:
  user: bob
  roles: prod-rw
  state: 1
  # [some fields were omitted for clarity]
  request_reason: "Prod is down and rollback failed"
  targets:
    - plugin: pagerduty

>       recipient: Teleport Alice

Reviewer: Can this be a list?

Each auth server should maintain a cache of all target rules to avoid having to request them for each request.

### Security

In case of conflict (teamA writes a routing rule matching access-requests from teamB), both targets
will be added to the access request. A misconfigured rule will result in a team being wrongly notified.
The main issue will be spam/noise but this should not lock out the rightful team to get notified by
the access request.

Misconfigured rules should not represent security issues unless the plugin does some kind of automatic approval.
This is the case for the pagerduty plugin which auto-approves if the requestor is on-call for the notified service.
In such situation, `access_routing_rules` should be considered as critical as `roles` themselves as they can lead
to automatic access escalation. Such risk is already present in the current implementation.

This risk can be mitigated by not granting the access-request edition rights to the plugin (at the price of a feature loss).

If the two targets are identical it is the plugin's responsibility to deduplicate before sending notifications
(to avoid sending duplicates and to ensure `PluginData` consistence).

> ### User Experience

Reviewer: Let's give a specific example here of how a scenario from https://github.com/gravitational/teleport-plugins/issues/597 could be implemented using access request routing rules.

Author: Thanks for the feedback, this helped me identify the required predicate helpers.

#### Complexity/Time to value

In most cases, a simple role_name check is performed and the users wants to set the plugin and the recipient.
This is provided by the simplified syntax:

```yaml
targets:
  - condition: 'equals(resource.spec.requested_role, "prod-rw")'
    recipient: "Teleport Alice"
    plugin: "pagerduty"
```

This syntax is equivalent to the pure-predicate one:

```yaml

> taregts:

Reviewer: ```suggestion
targets:
```

  - expression: >
      ifelse(
        equals(resource.spec.requested_role, "prod-rw"),
        pair("pagerduty", "Teleport Alice"),
        pair()
      )
```

> If users want to do complex operations and leverage all the predicate language power they can use pass the predicate code
> in the `expression` field. The predicate evaluation in `expression` field is expected to return a pair containing the plugin
> name and the recipient.b If one or both fields are empty, the target should not be set.

Reviewer: ```suggestion
If users want to do complex operations and leverage all the predicate language power they can use the predicate code
in the `expression` field. The predicate evaluation in `expression` field is expected to return a pair containing the plugin
name and the recipient. If one or both fields are empty, the target should not be set.
```

If `expression` is set, `condition`, `recipient` and `plugin` should not be used. Teleport must validate
this on resource creation and fail immediately instead of silently ignoring fields.

#### Combining with suggested_reviewers

The safest way would be for this new feature to not interact with suggested reviewers and let the plugins deal with it.

> Using `suggested_reviwer` implies the user knows which plugin will be used to send the access request. 

Reviewer: ```suggestion
Using `suggested_reviewer` implies the user knows which plugin will be used to send the access request. 
```

The string has to make sense and be allowed by the plugin. Today most plugins support email addresses
and reject other recipients, but somes are just ignoring the field (pagerduty). We might want to clarify
what can be entered in this field and how different plugins are reacting to it so user has insights on
what to type and what to expect.

Using `suggested_reviewer` with two plugins supporting it (e.g. Slack and MsTeams, or two Slack plugins on different workspaces)
can lead to the suggested reviewer being notified multiple times. One way to mitigate this would be to
offer a plugin configuration flag to ignore suggested reviewers (e.g. suggested reviewers are only
pinged on Slack).

#### JIT resource access request

> While `access_rouring_rule` should work with both role access requests and resource access requests,
> using it withresource access request can cause some friction for the admins writing the rules:

Reviewer: ```suggestion
While `access_routing_rule` should work with both role access requests and resource access requests,
using it with resource access request can cause some friction for the admins writing the rules:
```

> Only resource IDs are available through access requests: writing a rule to filter based on requested resources

Reviewer: FYI @nklaassen is adding some additional info to access requests to allow us to show actual resource names instead of IDs - Nic, could you link an RFD/PR (if there's any) here so we can see if it will work with the plugin routing rules?

Reviewer 2: @nklaassen I think the label-based routing is the best answer on how to support resource based access requests with this syntax.

requires knowing resources IDs. Rules with lists of IDs can be hard to read and maintain. For resource access requests routing,
admins would likely prefer to route the requests based on resources labels/tags.

A workaround would be to resolve the resources (either before evaluating the rule or through a new predicate function).

This is currently out of the scope of this RFD and could be implemented if the feature gets user traction.

### Plugin considerations

This change must be backward compatible:
- old plugins should ignore `targets` and get the recipients from the `role_to_recipient` map (or annotation for pagerduty);
- each plugin should expose an `honor_suggested_reviwers` configuration flag, true by default;
- each plugin should have a default name and listen to it by default in the `targets`, this name should be the same used to store plugin data;

> - admins can configure the plugin name through its configuration, names should always be prefixed by the plugin type (e.g. `slack-teanA`,

Reviewer: ```suggestion
- admins can configure the plugin name through its configuration, names should always be prefixed by the plugin type (e.g. `slack-teamA`,
```

  `msteams-teamB`) to limit conflicts between two plugins with the same names;

> - in DEBUG mode, each plugin should log if it is ignoring a target (because plugin name does not match for example) to ease troubleshooting; and

Reviewer: Nit: Just an implementation details but I'd probably just log at as Info. Should not clutter logs too much and it is a useful info to have imo, to be able to see whether what the plugin does with each request.

will be added to the access request. A misconfigured rule will result in a team being wrongly notified.
The main issue will be spam/noise but this should not lock out the rightful team to get notified by
the access request.

Misconfigured rules should not represent security issues unless the plugin does some kind of automatic approval.
This is the case for the pagerduty plugin which auto-approves if the requestor is on-call for the notified service.
In such situation, `access_request_routing_rules` should be considered as critical as `roles` themselves as they can lead
to automatic access escalation. Such risk is already present in the current implementation.

Such risk can be mitigated by not granting the access-request edition rights to the plugin (at the price of a feature loss).

If the two targets are identical it is the plugin's responsibility to deduplicate before sending notifications
(to avoid sending duplicates and to ensure `PluginData` unity).

### User Experience

#### Complexity/Time to value

In most cases, a simple role_name check is performed and the users wants to set the plugin and the recipients.
This is provided by the simplified syntax:

```yaml
targets:

>   - condition: 'resource.spec.roles.contains("prod-rw")'

Reviewer 2: this would be more robust if you treated this as `switch` behavior in go, there is no case fall-through, when case matches, it stops

Author: If I understand correctly, each `access_request_routing_rule` can provide at most one target, but users can rely on multiple `access_request_routing_rule` to add multiple targets.

While this approach should work with both role access request and resource access request, using it with
resource access request can cause some friction for admin writing the rules:

At the time of writing this proposal, only resource IDs are available through access requests: writing a rule to filter
based on requested resources requires knowing resources IDs. Rules with lists of IDs can be hard to read. For resource
access requests routing, admins are likely to want to route the requests based on resources labels/tags.
Roles are also available, but the requested role does not necessarily follow the principle of the least privilege.

When resource labels will be available, users will be able to match access requests routing rules against
the resource labels. The main challenge will be to find a way to express what the users wants in a usable way.
A list of each resource ID and their labels is not usable as-is, the list will need to be flattened in some way to
represent the access-requests labels. Two approaches are available:

> - returning the union of all labels: This appraoch allows users to request several non-related resources and have the\

Reviewer: ```suggestion
- returning the union of all labels: This approach allows users to request several non-related resources and have the
```

  access-request still reach all relevant teams. However this will cause a risk of privilege escalation through
  automatic approval, like the Pagerduty plugin. If this approach is chosen, we should ensure no automatic approval
  can happen on such requests.

> - returning the intersection of all labels: This approach is the sanest from a securoty point of view but can degrade

Reviewer: ```suggestion
- returning the intersection of all labels: This approach is the sanest from a security point of view but can degrade
```
